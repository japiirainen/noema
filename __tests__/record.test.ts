import { record, string } from "../src/index.js"
import { expectParseFailure, expectParseSuccess } from "./helpers.js"

describe("record", () => {
  test("should parse an simple records", () => {
    expectParseSuccess(record(string), {})
    expectParseSuccess(record(string), { a: "1" })
    expectParseSuccess(record(string), { a: "1", b: "2" })
  })

  test("should not parse non-objects", () => {
    expectParseFailure(record(string), "")
  })

  test("should not parse arrays", () => {
    expectParseFailure(record(string), [])
  })

  test("should not parse null", () => {
    expectParseFailure(record(string), null)
  })

  test("should only consider own properties", () => {
    const obj = Object.assign(Object.create({ a: "" }), { b: "2", c: "3" })
    expectParseSuccess(record(string), obj, { b: "2", c: "3" })
  })

  test("should not parse records with invalid values", () => {
    expectParseFailure(record(string), { a: "1", b: 2 })
  })

  test("path should nest with nested records", () => {
    expectParseFailure(record(record(string)), { a: { b: 1 } })
  })
})
