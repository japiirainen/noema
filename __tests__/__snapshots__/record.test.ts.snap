// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`record path should nest with nested records 1`] = `
[
  {
    "code": "invalid_type",
    "message": "Expected value to be a string",
    "path": "a.b",
    "value": 1,
  },
]
`;

exports[`record should not parse arrays 1`] = `
[
  {
    "code": "invalid_type",
    "message": "Expected value to be an object",
    "path": "",
    "value": [],
  },
]
`;

exports[`record should not parse non-objects 1`] = `
[
  {
    "code": "invalid_type",
    "message": "Expected value to be an object",
    "path": "",
    "value": "",
  },
]
`;

exports[`record should not parse null 1`] = `
[
  {
    "code": "invalid_type",
    "message": "Expected value to be an object",
    "path": "",
    "value": null,
  },
]
`;

exports[`record should not parse records with invalid keys 1`] = `
[
  {
    "code": "generic_issue",
    "message": "Not an a!",
    "path": "b",
    "value": "b",
  },
]
`;

exports[`record should not parse records with invalid values 1`] = `
[
  {
    "code": "invalid_type",
    "message": "Expected value to be a string",
    "path": "b",
    "value": 2,
  },
]
`;
